### ISSUE TITLE
Four templates declare every non-pointer argument int, so a float scalar arrives as its bit pattern

### ISSUE THREAD

**Issue by benkelaya** [bledden/triton-msl#11]
**Author:** Hocine Benkelaya (with NeuroBrix)

## The point that decides

This is already known in this file, and already fixed once. In
`_lower_strided_scalar_matmul` the scalar declaration reads the argument's
type:

```python
arg_decls.append(
    f"    device {triton_type_to_msl(arg.elem_type)}* "
    f"{arg.name}_buf [[buffer({i})]]")
```

with a comment above it naming this exact failure:

> *`alpha` and `beta`, which are fp32: declared int, the float bit pattern was
> reinterpreted as an integer and the epilogue scaled the accumulator by
> garbage (baddbmm, ULP 30686 against CUDA's 1).*

**Four other templates in the same file never got that fix** and still emit:

```python
arg_decls.append(f"    device int* {arg.name}_buf [[buffer({i})]]")
```

* `_lower_k_loop_dot_inline`
* `_lower_flip_template`
* `_lower_3d_reduce_template`
* `_lower_3d_argminmax_template`

The first is the simdgroup matmul template, so any `alpha * acc + beta * bias`
epilogue on that path is computed with the bit patterns.

## What is actually emitted

```c
kernel void addmm_kernel(
    device float* a_ptr    [[buffer(0)]],
    ...
    device int* alpha_buf  [[buffer(10)]],
    device int* beta_buf   [[buffer(11)]],
    ...
) {
    int alpha = alpha_buf[0];
    int beta  = beta_buf[0];
    ...
    c_ptr[gr * stride_cm + gc] =
        float((alpha * tg_st[sgitg * 64u + i] + beta * (float)bias_ptr[gc]));
}
```

`alpha` and `beta` are `float` in the kernel's signature. Declared `int`, the
bit pattern of `1.0f` — `0x3F800000` = 1065353216 — is what multiplies the
accumulator and the bias.

**It compiles, it runs, and it returns an array of the right shape.** No
warning, no fallback. The wrongness is a factor of a billion.

## Reproducer

```python
import torch, triton, triton.language as tl

@triton.jit
def scaled_mm(a_ptr, b_ptr, bias_ptr, c_ptr, M, N, K,
              stride_am, stride_bk, stride_cm, alpha, beta,
              BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr):
    pid = tl.program_id(0)
    nt = (N + BLOCK_N - 1) // BLOCK_N
    pm, pn = pid // nt, pid % nt
    offs_m = pm * BLOCK_M + tl.arange(0, BLOCK_M)
    offs_n = pn * BLOCK_N + tl.arange(0, BLOCK_N)
    offs_k = tl.arange(0, BLOCK_K)
    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)
    for k in range(0, K, BLOCK_K):
        a = tl.load(a_ptr + offs_m[:, None] * stride_am + (k + offs_k)[None, :])
        b = tl.load(b_ptr + (k + offs_k)[:, None] * stride_bk + offs_n[None, :])
        acc += tl.dot(a, b)
    bias = tl.load(bias_ptr + offs_n, mask=offs_n < N, other=0.0)
    acc = alpha * acc + beta * bias[None, :]        # <- the float scalars
    c = c_ptr + offs_m[:, None] * stride_cm + offs_n[None, :]
    tl.store(c, acc, mask=(offs_m[:, None] < M) & (offs_n[None, :] < N))

M = N = K = 64
a = torch.full((M, K), 0.01, dtype=torch.float32, device="mps")
b = torch.full((K, N), 0.02, dtype=torch.float32, device="mps")
bias = torch.full((N,), 0.5, dtype=torch.float32, device="mps")
c = torch.zeros(M, N, dtype=torch.float32, device="mps")
scaled_mm[((M // 32) * (N // 32),)](a, b, bias, c, M, N, K,
                                    a.stride(0), b.stride(0), c.stride(0), 1.0, 1.0,
                                    BLOCK_M=32, BLOCK_N=32, BLOCK_K=32)
torch.mps.synchronize()
want = K * 0.01 * 0.02 + 0.5
print("expected", want, "got", float(c[0, 0]), "ratio", float(c[0, 0]) / want)
```

```
expected 0.5128
got      5.46313e+08
ratio    1.06535e+09        <- 0x3F800000, the integer value of 1.0f
```

The ratio *is* the bit pattern, which is what makes this one self-evident.

## Scope, and what makes it hard to notice

It needs a float scalar argument, so a plain matmul is unaffected — only an
epilogue with a runtime `alpha`/`beta` (or any other non-integer scalar) sees
it. A caller whose scalars happen to be integers is also unaffected.

Worth a separate look: the same computation written FLAT, outside a K-loop, is
**refused** at compile time —

```
Refusing to emit silently-wrong output: matmul with a trailing compute epilogue
(bias/activation/fma/clamp) that the fused-epilogue template can't emit …
```

— while the K-loop form above compiles and returns the wrong numbers. I have
not established whether the guard and these four declarations share a cause; I
mention it only because a guard that fires on the smaller case and not on the
larger one is the kind of asymmetry worth a look from someone who knows the
pass.

## Suggested shape, not a patch

The fixed site is in the same file and reads `triton_type_to_msl(arg.elem_type)`.
Whether the four should simply call the same helper, or whether something else
depends on those buffers being `int`, is yours rather than mine.

## Environment

Apple M4 Pro, macOS 26.6.2 (Darwin 25.6.0), Python 3.12.0, triton 3.7.0,
Metal backend.

**Comment by bledden** [comment]
Thanks for pointing out that one template already handled this correctly while the others didn't. I applied a shared native-type mapping to scalar declarations and unpacking across six template sites, covering the K-loop dot, flip, 3-D reduction and 3-D argmin/argmax paths you named. Float scalars no longer arrive as their integer bit patterns on those paths, while integer arguments retain their own types.

I also addressed the workload, not just the declarations. The retained alpha/beta K-loop epilogue executed through the native pipeline with the expected 0.5128 result. The flat-versus-loop distinction you flagged is separate from scalar decoding: each route needs to prove that it preserves the post-dot computation. There is bounded non-looped dot/epilogue support as well as the recovered K-loop path, but removing a loop isn't a guarantee that an arbitrary shape takes a supported route. I haven't claimed a paired rerun of an unspecified flat rewrite of your exact example.

Supported scale/column-bias epilogues now include bounded rectangular tiles and fp16/bf16 storage with fp32 arithmetic. This isn't arbitrary epilogue support: other arithmetic, masks and address forms still need their own proof. Your original end-to-end check on 0.3.0 would be useful confirmation beyond the retained reproducer, and a flat form that still refuses is welcome as a separate capability example.

**Comment by benkelaya** [comment]
Thanks, and for checking the alpha/beta epilogue end to end. If a flat form still refuses on 0.3.0, we'll open it as a separate capability example.
